Mathlib/NumberTheory/NumberField/Basic.lean

```diff
@@ -64,35 +64,76 @@ instance [NumberField L] [Algebra K L] : FiniteDimensional K L :=
   Module.Finite.of_restrictScalars_finite ℚ K L
 
 /-- The ring of integers (or number ring) corresponding to a number field
-is the integral closure of ℤ in the number field. -/
-def ringOfIntegers :=
+is the integral closure of ℤ in the number field.
+
+This is defined as its own type, rather than a `Subalgebra`, for performance reasons:
+looking for instances of the form `SMul (RingOfIntegers _) (RingOfIntegers _)` makes
+much more effective use of the discrimination tree than instances of the form
+`SMul (Subtype _) (Subtype _)`.
+The drawback is we have to copy over instances manually.
+-/
+def RingOfIntegers : Type _ :=
   integralClosure ℤ K
-#align number_field.ring_of_integers NumberField.ringOfIntegers
+#align number_field.ring_of_integers NumberField.RingOfIntegers
 
-@[inherit_doc] scoped notation "𝓞" => NumberField.ringOfIntegers
+@[inherit_doc] scoped notation "𝓞" => NumberField.RingOfIntegers
 
-theorem mem_ringOfIntegers (x : K) : x ∈ 𝓞 K ↔ IsIntegral ℤ x :=
-  Iff.rfl
-#align number_field.mem_ring_of_integers NumberField.mem_ringOfIntegers
+#noalign number_field.mem_ring_of_integers
 
-theorem isIntegral_of_mem_ringOfIntegers {K : Type*} [Field K] {x : K} (hx : x ∈ 𝓞 K) :
-    IsIntegral ℤ (⟨x, hx⟩ : 𝓞 K) := by
-  obtain ⟨P, hPm, hP⟩ := hx
-  refine' ⟨P, hPm, _⟩
-  rw [← Polynomial.aeval_def, ← Subalgebra.coe_eq_zero, Polynomial.aeval_subalgebra_coe,
-    Polynomial.aeval_def, Subtype.coe_mk, hP]
-#align number_field.is_integral_of_mem_ring_of_integers NumberField.isIntegral_of_mem_ringOfIntegers
+namespace RingOfIntegers
+
+instance : CommRing (𝓞 K) :=
+  inferInstanceAs (CommRing (integralClosure _ _))
```

Comment:
Let me try.

Reply:
`abbrev` is much slower on my computer. I didn't investigate, but I think that, since the definition is now reducible, Lean knows that `𝓞 K` is the coercion of a subalgebra, so it still goes through the instances we want to avoid.

Reply:
Indeed, making an `abbrev` would undo the point of this change, to hide the coercion arrow behind the definition.